test/base/SpecBase.scala

```diff
@@ -16,37 +16,69 @@
 import controllers.actions._
 import models.UserAnswers
+import org.joda.time.LocalDate
 import org.scalatest.concurrent.{IntegrationPatience, ScalaFutures}
 import org.scalatest.freespec.AnyFreeSpec
 import org.scalatest.matchers.must.Matchers
 import org.scalatest.{OptionValues, TryValues}
 import play.api.Application
-import play.api.i18n.{Messages, MessagesApi}
+import play.api.i18n.{Lang, Messages, MessagesApi}
 import play.api.inject.bind
 import play.api.inject.guice.GuiceApplicationBuilder
+import play.api.mvc.AnyContentAsEmpty
 import play.api.test.FakeRequest
 
-trait SpecBase
-  extends AnyFreeSpec
-    with Matchers
-    with TryValues
-    with OptionValues
-    with ScalaFutures
-    with IntegrationPatience {
+import scala.concurrent.ExecutionContext
+
+trait SpecBase extends AnyFreeSpec with Matchers with TryValues with OptionValues with ScalaFutures with IntegrationPatience {
 
   val userAnswersId: String = "id"
+  val taxYear: Int          = LocalDate.now().getYear
+  val enLang                = Lang("en-EN")
+  val cyLang                = Lang("cy-CY")
+
+  implicit val ec                   = ExecutionContext.global
+  def emptyUserAnswers: UserAnswers = UserAnswers(userAnswersId)
+
+  def messages(app: Application, isWelsh: Boolean = false): Messages =
+    if (isWelsh) {
+      app.injector.instanceOf[MessagesApi].preferred(Seq(Lang("cy")))
+    } else {
+      app.injector.instanceOf[MessagesApi].preferred(FakeRequest().withHeaders())
+    }
 
-  def emptyUserAnswers : UserAnswers = UserAnswers(userAnswersId)
+  protected def isWelshToString(isWelsh: Boolean): String = if (isWelsh) "Welsh" else "English"
 
-  def messages(app: Application): Messages = app.injector.instanceOf[MessagesApi].preferred(FakeRequest())
+  protected def isAgentToString(isAgent: Boolean): String = if (isAgent) "agent" else "individual"
+
+  protected def applicationBuilder(userAnswers: Option[UserAnswers] = None, isAgent: Boolean = false): GuiceApplicationBuilder = {
+    val fakeIdentifierAction = {
+      if (isAgent) {
+        bind[IdentifierAction].to[FakeAgentIdentifierAction]
+      } else {
+        bind[IdentifierAction].to[FakeIndividualIdentifierAction]
+      }
+    }
 
-  protected def applicationBuilder(userAnswers: Option[UserAnswers] = None): GuiceApplicationBuilder =
     new GuiceApplicationBuilder()
       .overrides(
         bind[DataRequiredAction].to[DataRequiredActionImpl],
-        bind[IdentifierAction].to[FakeIdentifierAction],
+        fakeIdentifierAction,
         bind[DataRetrievalAction].toInstance(new FakeDataRetrievalAction(userAnswers))
       )
+  }
+
+  protected def buildRequest(httpType: String, requestRoute: String, isAgent: Boolean): FakeRequest[AnyContentAsEmpty.type] = {
+    if (isAgent) {
+      FakeRequest(httpType, requestRoute)
+        .withSession(SessionValues.CLIENT_MTDITID -> aNoddyUser.mtditid, SessionValues.CLIENT_NINO -> aNoddyUser.nino)
```

Comment:
This seems strange - if we are an agent then we are using the individual credentials? Do we need this differentiator here? In the controller specs, we are acting as an individual or agent during binding in the `applicationBuilder` method. Could we just send a `FakeRequest(httpType, requestRoute)` - it looks like we have done this in some of the tests but not all. Can we do it in all?

Reply:
You're right, this was part of an old system for agent/individual Edem and I were making but is no longer needed. Have removed, thanks